File: dijkstras/path-with-minimum-effort/path-with-minimum-effort.java
class Solution {
  public static class Cell {
    int effort;
    int i;
    int j;

    Cell(int i, int j, int effort) {
      this.i = i;
      this.j = j;
      this.effort = effort;
    }
  }

  public int minimumEffortPath(int[][] heights) {
    int h = heights.length;
    int w = heights[0].length;

    int[][] directions = new int[][] { { 0, -1 }, { 0, 1 }, { -1, 0 }, { 1, 0 } };

    int[][] efforts = new int[h][w];

    for (int[] effort : efforts) {
      Arrays.fill(effort, Integer.MAX_VALUE);
    }

    Comparator<Cell> comp = new Comparator<>() {
      public int compare(Cell i1, Cell i2) {
        return i1.effort - i2.effort;
      }
    };

    PriorityQueue<Cell> queue = new PriorityQueue<>(1, comp);

    boolean[][] visited = new boolean[h][w];

    int i = 0;
    int j = 0;

    efforts[i][j] = 0;

    queue.offer(new Cell(0, 0, 0));

    while (!queue.isEmpty()) {
      Cell next = queue.poll();
      i = next.i;
      j = next.j;

      if (visited[i][j]) {
        continue;
      }

      if (i == h - 1 && j == w - 1) {
        return next.effort;
      }

      visited[i][j] = true;

      int current = heights[i][j];
      int currentEffort = efforts[i][j];

      for (int[] dir : directions) {
        int newi = i + dir[0];
        int newj = j + dir[1];

        if (newi < 0 || newj < 0 || newi == h || newj == w) {
          continue;
        }

        int diff = Math.abs(current - heights[newi][newj]);

        int maxEffort = Math.max(diff, currentEffort);

        if (maxEffort < efforts[newi][newj]) {
          efforts[newi][newj] = maxEffort;
          queue.offer(new Cell(newi, newj, maxEffort));
        }
      }
    }

    return efforts[h - 1][w - 1];
  }
}

File: dijkstras/path-with-minimum-effort/path-with-minimum-effort.js
class Cell {
  constructor(i, j, effort) {
    this.i = i;
    this.j = j;
    this.effort = effort;
  }
}

/**
 * @param {number[][]} heights
 * @return {number}
 */
function minimumEffortPath(heights) {
  const h = heights.length;
  const w = heights[0].length;

  const directions = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  const efforts = new Array(h).fill(null).map(() => new Array(w).fill(Number.MAX_SAFE_INTEGER));
  const queue = new MinPriorityQueue({ priority: (bid) => bid.effort });

  const visited = new Array(h).fill(null).map(() => new Array(w).fill(false));

  let i = 0;
  let j = 0;

  efforts[i][j] = 0;

  queue.enqueue(new Cell(0, 0, 0));

  while (!queue.isEmpty()) {
    const next = queue.dequeue().element;
    const i = next.i;
    const j = next.j;

    if (visited[i][j]) {
      continue;
    }

    if (i == h - 1 && j == w - 1) {
      return next.effort;
    }

    visited[i][j] = true;

    let current = heights[i][j];
    let currentEffort = efforts[i][j];

    for (let dir of directions) {
      let newi = i + dir[0];
      let newj = j + dir[1];

      if (newi < 0 || newj < 0 || newi == h || newj == w) {
        continue;
      }

      let diff = Math.abs(current - heights[newi][newj]);

      let maxEffort = Math.max(diff, currentEffort);

      if (maxEffort < efforts[newi][newj]) {
        efforts[newi][newj] = maxEffort;
        queue.enqueue(new Cell(newi, newj, maxEffort));
      }
    }
  }

  return efforts[h - 1][w - 1];
}
